src/control/lib/netdetect/netdetect.go

```diff
@@ -887,37 +993,70 @@ func ScanFabric(provider string) ([]FabricScan, error) {
 	log.Debugf("initDeviceScan completed.  Depth %d, numObj %d, systemDeviceNames %v, hwlocDeviceNames %v",
 		deviceScanCfg.depth, deviceScanCfg.numObj, deviceScanCfg.systemDeviceNames, deviceScanCfg.hwlocDeviceNames)
 
+	var hfiDeviceCount int
+	for _, device := range deviceScanCfg.hwlocDeviceNames {
+		if strings.Contains(device, "hfi1_") {
+			hfiDeviceCount++
+			log.Debugf("There are %d hfi1 devices in the system", hfiDeviceCount)
+		}
+	}
+
 	for ; fi != nil; fi = fi.next {
 		if fi.domain_attr == nil || fi.domain_attr.name == nil || fi.fabric_attr == nil || fi.fabric_attr.prov_name == nil {
 			continue
 		}
-
 		deviceScanCfg.targetDevice = C.GoString(fi.domain_attr.name)
-		deviceAffinity, err := GetAffinityForDevice(deviceScanCfg)
-		if err != nil {
-			log.Debugf("Error from GetAffinityForDevice: %v", err)
-			continue
-		}
 
-		// Convert the libfabric provider list to a Mercury compatible provider list
-		mercuryProviderList, err := convertLibFabricToMercury(C.GoString(fi.fabric_attr.prov_name))
+		// Implements a workaround to handle the current psm2 provider behavior
+		// that reports fi.domain_attr.name as "psm2" instead of an actual device
+		// name like "hfi1_0".
+		//
+		// The work around behaves in two different ways depending on the data available.
+		// When the value of fi.src_addr.unit is 0 or greater, it is used to explicitly
+		// define an hfi1_ unit by appending that number to the "hfi1_" string.
+		//
+		// When the value of fi.src_addr.unit is -1, it means that all hfi1_x units in the system
+		// are supported by the psm2 provider.  In this case, we need to know how many hfi1_x units
+		// there are, and create a fabric scan entry for each one.  By querying the device names from
+		// hwloc, we can count the number of devices matching "hfi1_" and create a device scan
+		// entry for each one.
+		if C.GoString(fi.fabric_attr.prov_name) == "psm2" {
+			if strings.Contains(deviceScanCfg.targetDevice, "psm2") {
+				log.Debugf("psm2 provider and psm2 device found.")
+				hfiUnit := C.getHFIUnit(fi.src_addr)
+				switch hfiUnit {
+				case allHFIUsed:
+					for deviceID := 0; deviceID < hfiDeviceCount; deviceID++ {
+						deviceScanCfg.targetDevice = fmt.Sprintf("hfi1_%d", deviceID)
+						scanResults, err := createFabricScanEntry(deviceScanCfg, C.GoString(fi.fabric_attr.prov_name), devCount)
+						if err != nil {
+							// errors from createFabricScanEntry are non fatal errors
+							continue
+						}
+						results := scanResults.String()
+						if _, found := resultsMap[results]; !found {
+							resultsMap[results] = struct{}{}
+							log.Debugf("\n%s", results)
+							ScanResults = append(ScanResults, scanResults)
+							devCount++
+						}
```

Comment:
Looks like this is repeated below - another case where cleaning up into a helper method would be nice. What does it mean that we do all of this again?

Reply:
I've consolidated it.